azure-ts-msi-keyvault-rbac/index.ts

```diff
@@ -0,0 +1,170 @@
+import * as pulumi from "@pulumi/pulumi";
+import * as azure from "@pulumi/azure";
+import { execSync} from "child_process";
+
+// Create a resource group
+const resourceGroup = new azure.core.ResourceGroup("resourceGroup", {
+    location: "West US",
+});
+
+// Create a storage account for Blobs
+const storageAccount = new azure.storage.Account("storage", {
+    resourceGroupName: resourceGroup.name,
+    accountReplicationType: "LRS",
+    accountTier: "Standard",
+});
+
+// The container to put our files into
+const storageContainer = new azure.storage.Container("files", {
+    resourceGroupName: resourceGroup.name,
+    storageAccountName: storageAccount.name,
+    containerAccessType: "private",
+});
+
+// Azure SQL Server that we want to access from the application
+const sqlServer = new azure.sql.SqlServer("sqlserver", {
+    resourceGroupName: resourceGroup.name,
+
+    administratorLogin: "manualadmin",
+    administratorLoginPassword: "WeWontUseTh1sPwd",
+    version: "12.0",
+});
+
+// Azure SQL Database that we want to access from the application
+const database = new azure.sql.Database("sqldb", {
+    resourceGroupName: resourceGroup.name,
+    serverName: sqlServer.name,
+    requestedServiceObjectiveName: "S0"
+});
+
+// The connection string that has no credentials in it: authertication will come through MSI
+const connectionString = pulumi.all([sqlServer.name, database.name]).apply(([server, db]) => 
+    `Server=tcp:${server}.database.windows.net;Database=${db};`);
+
+// A file in Blob Storage that we want to access from the application
+const textBlob = new azure.storage.Blob("text", {
+    resourceGroupName: resourceGroup.name,
+    storageAccountName: storageAccount.name,
+    storageContainerName: storageContainer.name,
+    type: "block",
+    source: "./README.md",
+});
+
+// A plan to host the App Service
+const appServicePlan = new azure.appservice.Plan("asp", {
+    resourceGroupName: resourceGroup.name,
+    kind: "App",
+    sku: {
+        tier: "Basic",
+        size: "B1",
+    },
+});
+
+// ASP.NET deployment package
+const blob = new azure.storage.ZipBlob("zip", {
+    resourceGroupName: resourceGroup.name,
+    storageAccountName: storageAccount.name,
+    storageContainerName: storageContainer.name,
+    type: "block",
+
+    content: new pulumi.asset.FileArchive("./webapp/bin/Debug/netcoreapp2.2/publish")
+});
+
+const clientConfig = pulumi.output(azure.core.getClientConfig({}));
+const tenantId = clientConfig.apply(c => c.tenantId);
+
+// Determine the ID of the current user or service principal
+const currentPrincipal = clientConfig.apply(c => 
+    c.servicePrincipalObjectId !== "" 
+    // Current, only service principal ID is available in the context
+    ? c.servicePrincipalObjectId
+    // If logged in with a user, find their ID via Azure CLI
+    : <string>JSON.parse(execSync("az ad signed-in-user show --query objectId").toString()));
```

Comment:
Interesting.  This is not going to be robust in 100% of cases, but I kind of like it to highlight how to pragmatically do "awkward" things like this.

Reply:
Yes, this is based on https://github.com/terraform-providers/terraform-provider-azurerm/issues/3234 (I'm adding the link to this issue into comments)

Reply:
fwiw, this code just looks hard to read.  Consider just tweaking into:

```ts
const currentPrincipal = clientConfig.apply(c =>
    // Currently, only service principal ID is available in the context.  If we are provided the principle in the config, then
    // just use it.  Otherwise, if logged in with a user, find their ID via Azure CLI.
    // see https://github.com/terraform-providers/terraform-provider-azurerm/issues/3234
    c.servicePrincipalObjectId !== ""
        ? c.servicePrincipalObjectId
        : <string>JSON.parse(execSync("az ad signed-in-user show --query objectId").toString()));
```

Reply:
Tweaked, thanks!